src/localize-router.service.ts

```diff
@@ -43,8 +44,32 @@ export class LocalizeRouterService {
       let rootSnapshot: ActivatedRouteSnapshot = this.router.routerState.snapshot.root;
 
       this.parser.translateRoutes(lang).subscribe(() => {
-        const url = this.traverseRouteSnapshot(rootSnapshot);
-        
+        let url = this.traverseRouteSnapshot(rootSnapshot);
+
+        if (!this.settings.alwaysSetPrefix) {
+          var urlSegments = url.split('/');
+          //If the default language has no prefix make sure to remove and add it when necessary
+          if (this.parser.currentLang === this.parser.defaultLang) {
+            //Remove the language prefix from url when current language is the default language
+            if (url.indexOf(this.parser.currentLang + '/') === -1) {
+              //Remove the current aka default language prefix from the url
+              urlSegments = urlSegments.splice(urlSegments.indexOf(this.parser.currentLang), 1);
+              url = urlSegments.join('/');
+            }
+          } else if (this.parser.currentLang !== this.parser.defaultLang) {
+            //When coming from a default language it's possible that the url doesn't contain the language, make sure it does.
+            if (url.indexOf(this.parser.currentLang + '/') === -1) {
+              //If the url starts with a slash make sure to keep it.
+              if (urlSegments[0] === '') {
+                urlSegments.splice(1, 0, this.parser.currentLang);
```

Comment:
This could be simplified by:
const injectionIndex = urlSegments[0] ? 0 : 1;
urlSegments.splice(injectionIndex, 0, this.parser.currentLang);
url = urlSegments.join('/');

Ideally, I would not use splice as it's rather expensive operation and use pure functions instead like `slice` and `concat`